syntax = "proto3";
package streetlevel;

import "Shared.proto";

message PlaceRequest {
    AnalyticMetadata analytic_metadata = 1;
    ClientMetadata client_metadata = 2;
    repeated string display_language = 3;
    string spoken_language = 4;
    repeated ComponentInfo request_component = 5;
    bool suppress_results_requiring_attribution = 6;
    RequestType request_type = 7;
    PlaceRequestParameters place_request_parameters = 8;
    string display_region = 9;
    repeated ComponentInfo auxiliary_tier_requested_component = 11;
    //PrivacyMetadata privacy_metadata = 12;

    message AnalyticMetadata {
        string app_identifier = 1;
        string app_major_version = 2;
        string app_minor_version = 3;
        string hardware_model = 4;
        string os_version = 5;
        bool is_internal_tool = 6;
        bool is_internal_install = 7;
        SessionID session_id = 8;
        int64 sequence_number = 9;
        int64 request_source = 10;
        bool is_from_api = 12;
        GeoServiceTag service_tag = 13;
        string product_name = 14;
        int32 logged_ab_experiment = 15;
        int32 ab_assign_info = 16;
        LocalTime request_time = 17;
        uint64 relative_timestamp = 18;
        bool is_siri_original_request = 19;
        uint64 route_stop_count = 20;
    }

    message SessionID {
        uint64 unknown1 = 1;
        uint64 unknown2 = 2;
    }

    message GeoServiceTag {
        ServiceType service_type = 1;
        string tag = 2;
    }

    message LocalTime {
        uint64 time_rounded_to_hour = 1;
        int32 timezone_offset_from_gmt_in_hours = 2;
        string human_readable = 2000;
    }

    message ClientMetadata {
        string device_country_code = 1;
        string device_display_language = 3;
        string device_keyboard_language = 4;
        string device_spoken_language = 5;
        uint32 time_since_map_entered_foreground = 8;
        //AdditionalEnabledMarkets additional_enabled_markets = 9;
        bool result_list_attribution_support = 10;
        uint32 hour_of_day = 11;
        uint32 day_of_week = 12;
        //GeoLocation device_extended_location = 13;
        //GeoLocation device_historical_location = 15;
        int32 required_version = 16;
        string debug_api_key = 17;
        repeated int32 known_client_resolved_type_deprecated = 18;
        bool enable_preflight_venues = 19;
        repeated int32 known_client_resolved_type = 20;
        string business_chat_preflight_identifier = 21;
        int32 client_revision = 22;
        LocalizationCapabilities localization_capabilities = 23;
        bool siri_user_consents_for_analysis = 24;
        repeated MapsResultType supported_maps_result_type = 26;
        //MapsSuggestionsTouristInfo maps_suggestions_tourist_info = 27;
        repeated int32 client_revisions = 28;
        string preferred_display_currency_symbol = 29;
        string display_region = 30;
        bool support_enrichment = 31;
    }

    message LocalizationCapabilities {
        repeated SupportedPhoneticType supported_phonetic_type = 1;
        bool supports_localized_traffic_control_icons = 2;
    }

    message PlaceRequestParameters {
        //SearchParameters search_parameters = 1;
        //GeocodingParameters geocoding_parameters = 2;
        //CanonicalLocationSearchParameters canonical_location_search_parameters = 3;
        ReverseGeocodingParameters reverse_geocoding_parameters = 4;
        message ReverseGeocodingParameters {
            LatLng location = 1;
            int32 place_type_limit = 2;
            bool preserve_original_location = 3;
            Location extended_location = 4;
        }
        //PlaceLookupParameters place_lookup_parameters = 5;
        //MerchantLookupParameters merchant_lookup_parameters = 6;
        //PlaceRefinementParameters place_refinement_parameters = 7;
        //SiriSearchParameters siri_search_parameters = 8;
        //LocationDirectedSearchParameters location_directed_search_parameters = 9;
        //AutocompleteParameters autocomplete_parameters = 10;
        //etc. etc. etc.
        //TODO Add all this stuff
    }

    message Location {
        LatLng lat_lng = 1;
        LocationType type = 2;
        double timestamp = 3;
        int32 altitude = 4;
        double horizontal_accuracy = 5;
        double vertical_accuracy = 6;
        double speed = 7;
        double heading = 8;
        double course = 9;
        int32 level_ordinal = 10;
        //etc. etc. etc. TODO
    }
}

message ComponentInfo {
    int32 type = 1;
    int32 start_index = 2;
    int32 count = 3;
    ComponentFilter component_filter = 5;
    bool include_source = 6;
    int32 urgency = 7;
}

message ComponentFilter {
    // TODO All these fucking filter messages. There's dozens of these fucking things
    AddressObjectFilter address_object_filter = 31;
    message AddressObjectFilter {
        repeated string library_version = 1;
        bool enable_location = 2;
    }
}

enum ServiceType {
    SERVICE_UNKNOWN = 0;
    REV_GEO = 1;
    FWD_GEO = 2;
    SEARCH = 3;
    ROUTING = 4;
    LOCATION_SERVICE = 5;
    COLLECTION = 6;
    CELL_WIFI_HARVEST = 7;
    TRACE_HARVEST = 8;
    TRAFFIC_TILES = 9;
    RASTER = 10;
    AUTO_COMPLETE = 11;
    DASHBOARD_TEST = 12;
    VECTOR = 13;
    ADDR_PTS_HARVEST = 14;
    DISPATCHER = 15;
    PLACE_LOOKUP = 16;
    PARSEC = 17;
    GEO_SERVICES_FRAMEWORK = 18;
    SIRI = 19;
    MERCHANT_LOOKUP = 20;
    GEO_SEARCH = 21;
    TRAFFIC_ETA = 22;
    COMMUNITY = 23;
    LOCUS = 24;
    AUTONAVI_FRONTDOOR = 25;
    COMMUNITY_NON_PII = 26;
    DISPATCHER_ENRICHMENT = 27;
}

enum SupportedPhoneticType {
    NUANCE_LHPLUS = 0;
    UNABBREVIATED = 1;
    XSAMPA = 2;
}

enum LocationType {
    GPS = 0;
    WIFI = 1;
    CELL = 2;
    IP_ADDRESS = 3;
}
